beacon-chain/sync/data_columns.go

```diff
@@ -0,0 +1,312 @@
+) ([]blocks.RODataColumn, error) {
+	if len(dataColumns) == 0 {
+		return nil, nil
+	}
+
+	// Assemble the peers who can provide the needed data columns.
+	dataColumnsByAdmissiblePeer, _, _, err := p2p.AdmissiblePeersForDataColumns(peers, dataColumns)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get admissible peers for data columns")
+	}
+
+	badPeers := make(map[peer.ID]bool)
+	sidecars := make([]blocks.RODataColumn, 0, len(dataColumns))
+	remainingColumns := make(map[uint64]bool, len(dataColumns))
+	for col := range dataColumns {
+		remainingColumns[col] = true
+	}
+
+	for len(peers) > len(badPeers) && len(remainingColumns) > 0 {
+		// Filter out bad peers from the admissible peers
+		filteredDataColumnsByAdmissiblePeer := make(map[peer.ID]map[uint64]bool)
+		for p, cols := range dataColumnsByAdmissiblePeer {
+			if !badPeers[p] {
+				filteredDataColumnsByAdmissiblePeer[p] = cols
+			}
+		}
+
+		// Try to select peers excluding bad peers
+		peersToFetchFrom, err := SelectPeersToFetchDataColumnsFrom(remainingColumns, filteredDataColumnsByAdmissiblePeer)
+		if err != nil {
+			// Return an error if some columns are unavailable from the filtered set
+			// of peers. Filtering out bad peers can make columns unavailable, and
+			// when that happens, the caller needs to know about it.
+			return nil, errors.Wrap(err, "couldn't select peers to fetch data columns from")
+		}
+
+		// Request the data columns from each peer
+		successfulColumns := make(map[uint64]bool)
+		for peer, dataColumns := range peersToFetchFrom {
+			request, err := RequestsForDataColumnsByRoot(blkRoot, dataColumns)
+			if err != nil {
+				log.WithError(err).Debug("Failed to build request for data columns")
+				continue
+			}
+
+			peerSidecars, err := SendDataColumnSidecarsByRootRequest(ctx, clock, p2p, peer, ctxMap, &request)
+			if err != nil {
+				// Mark this peer as bad since it failed to respond correctly
+				badPeers[peer] = true
+				log.WithFields(logrus.Fields{
+					"peer":      peer.String(),
+					"blockRoot": fmt.Sprintf("%#x", blkRoot),
+					"error":     err.Error(),
+				}).Debug("Failed to request data columns from peer")
+				continue
+			}
+
+			// Mark columns as successful
+			for _, sidecar := range peerSidecars {
+				colIndex := sidecar.ColumnIndex
+				successfulColumns[colIndex] = true
+			}
+
+			for _, colIndex := range dataColumns {
+				if !successfulColumns[colIndex] {
+					// Mark peer as bad if any requested column wasn't successful
+					badPeers[peer] = true
+					log.WithFields(logrus.Fields{
+						"peer":          peer.String(),
+						"missingColumn": colIndex,
+					}).Debug("Peer failed to return requested data column")
+					break
+				}
+			}
+
+			sidecars = append(sidecars, peerSidecars...)
+		}
+
+		// Update remaining columns for the next retry
+		for col := range successfulColumns {
+			delete(remainingColumns, col)
+		}
+
+		if len(remainingColumns) == 0 {
+			break
```

Comment:
Instead of breaking here (which is the happy path), I think it's better to put all the code after 
```go
// Validate the received sidecars
```

So, exiting the for loop is the unhappy path, and does not need to test any more
```go
if len(remainingColumns) > 0
```

Reply:
That check was leftover from a previous iteration and isn't needed at all anymore. But I don't understand what you were suggesting about moving code around. I just removed it for now but let me know if there was something else you were looking for!

Reply:
https://github.com/prysmaticlabs/prysm/pull/14977/commits/34c985870636352360f86df02149357e180cd441